### PR TITLE
[com_fields] Handling single 0 selections properly in checkboxes

### PR DESCRIPTION
**Pull request by laoneo**
Pull Request for Issue #12810.

### Summary of Changes
Adds a callback functionality for form fields to prepare the value of a custom field before the value is passed to the form.

### Testing Instructions
- Create a custom field of the type checkboxes and assign two options, one of them with a 0 value. Eg: 
image.
- Select only the 0 value ("No" in my example) and save.
- Reopen the form and check the value.
- Check the item in frontend.

### Expected result
When editing the article the 0 checkbox should be selected. On the frontend "No" should be displayed.

### Actual result
No checkbox is selected in the form and nothing is displayed when viewing an article.

### CODE REVIEW

**Comment by Bakual** [comment]
So we need a prepareValue method to for this FormField?

**Comment by laoneo** [comment]
??

**Comment by Bakual** [comment]
This code basically is only needed for the checkboxes field, right? And only to be used by com_fields?
Looks a bit strange to me then.
Since the checkboxes field works fine with JForm, I'd expect a fix in com_fields, not in the JFormField.

**Comment by laoneo** [comment]
It allows a field to influence the value. If some more advanced fields will be created in the future they can then influence how a value is set on the form.
But right now it is only used in custom fields. What I want to prevent is to use field specific code in com_fields itself.

**Comment by Bakual** [comment]
> they can then influence how a value is set on the form

Isn't that already done with getInput? :smile: The appearance in a form is in full control of the formfield already.
The issue is only in com_fields in how it tries to guess if it should show the field value or not (I think). But as written in the issue I haven't found where the difference comes from. My guess was it's related to "forceMultiple" being true in the checkboxes field.

> What I want to prevent is to use field specific code in com_fields itself.

Good idea :+1:

**Comment by laoneo** [comment]
The problem lies here https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_fields/models/field.php#L581 where no array is set when it is a single row value.

**Comment by Bakual** [comment]
And what does JForm differently that we get an array even with only one value? That's what I haven't found out yet. Something must be different and I think it's the cause for this issue.

**Comment by laoneo** [comment]
I'v tested it adding a checkboxs field to the attribs fieldset. There it is stored as array when it get's json encoded.

**Comment by Bakual** [comment]
I see, if a field has the attribute "multiple" set (also list fields), then the value is stored as array in the JSON string.
Most fields however can process strings as well since the attribute is optional. In the case of the checkboxes it (rightfully) always expects an array since it has forceMultiple property set to true.

I don't think we can check the multiple property correct since it's set during runtime of the field based on the forceMultiple property. Or do you see a way for that? Then it would be simple.

If we have to adjust the checkboxes field anyway I'd rather just cast the value to array and be done.

**Comment by laoneo** [comment]
I don't really understand what this forceMultiple property does. Just had a quick look on it.

**Comment by Bakual** [comment]
It ensures `$this->multiple` is always true and thus the value is always saved as an array 😄

**Comment by laoneo** [comment]
forceMultiple did the trick

File: administrator/components/com_fields/helpers/fields.php [hunk 1]
```diff
@@ -472,6 +472,14 @@ public static function prepareForm($context, JForm $form, $data)
 			{
 				$value = $model->getFieldValue($field->id, $field->context, $data->id);
 
+				// Creating the XML form data
+				$formField = $form->getField($field->alias, 'params');
+
+				if ($formField && !is_array($value) && $value !== '' && $value !== null && $formField->forceMultiple)
```
Comment (Bakual) [inline administrator/components/com_fields/helpers/fields.php]: Maybe put the formField fetching into an `if (!is_array($value) && $value !== '' && $value !== null)` clause? No need to do something if we have an empty value or the value is already an array.

**Comment by Bakual** [comment]
Looks much better this way. Thanks!

File: administrator/components/com_fields/helpers/fields.php (hunk 1 above)
Comment (laoneo) [inline administrator/components/com_fields/helpers/fields.php]: It's already an if in a foor loop within an if, don't want to add another `if ($formField)`

Comment (Bakual) [inline administrator/components/com_fields/helpers/fields.php]: You can get rid of the first if easily :smile: 
````
if (!isset($data->id) || !$data->id)
{
	return true;
}

foreach ($fields as $field)
{
	....
}
````
Also I would move the forceMultiple check a bit down after the check for null, one thing less to check then.

Comment (laoneo) [inline administrator/components/com_fields/helpers/fields.php]: `!empty($data->id)` will throw a notice when the id is not set

**Comment by laoneo** [comment]
@Bakual could you test it?

**Comment by Bakual** [comment]
I have tested this item :white_check_mark: successfully on 94be45f90095fac585c0b6c01eeab6d0c80eb859<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12886">issues.joomla.org/tracker/joomla-cms/12886</a>.</sub>

**Comment by Bakual** [comment]
Thanks for the reminder. I wanted to test it yesterday but forgot it :smile: 

The issue #12812 still exists, but is probably unrelated to the issue here.

**Comment by laoneo** [comment]
Yes, that's different

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 94be45f90095fac585c0b6c01eeab6d0c80eb859<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12886">issues.joomla.org/tracker/joomla-cms/12886</a>.</sub>

**Comment by jeckodevelopment** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12886">issues.joomla.org/tracker/joomla-cms/12886</a>.</sub>

**Comment by rdeutz** [comment]
@laoneo could you check the conflicts, thanks

**Comment by laoneo** [comment]
Done

**Comment by laoneo** [comment]
Thanks.

**Pull request merged**